src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,356 @@
+        dstd = os.path.join(conf.pth, upa)
+        os.makedirs(dstd, exist_ok=True)
+
+        dst = os.path.join(dstd, f"{upa}.fa")
+        # Hard link .fa file from job_dir to output_dir in WS
+        os.link(cfn, dst)
+
+        metafile = os.path.join(dstd, f"{upa}.meta")
+        # save meta file with relevant object_info
+        with open(metafile, "w", encoding="utf8") as json_file:
+            json.dump(_process_object_info(obj_info), json_file, indent=2)
+
+        print("Completed %s" % (upa))
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="PROTOTYPE - Download genome files from the workspace service (WSS)."
+    )
+
+    required = parser.add_argument_group("required named arguments")
+    optional = parser.add_argument_group("optional arguments")
+
+    # Required flag argument
+    required.add_argument(
+        "--workspace_id",
+        required=True,
+        type=int,
+        help="Workspace addressed by the permanent ID",
+    )
+
+    # Optional argument
+    optional.add_argument(
+        "--root_dir",
+        type=str,
+        default=loader_common_names.ROOT_DIR,
+        help="Root directory.",
+    )
+    optional.add_argument(
+        "--kb_base_url",
+        type=str,
+        default=loader_common_names.KB_BASE_URL,
+        help="KBase base URL, defaulting to prod",
+    )
+    optional.add_argument(
+        "--workers",
+        type=int,
+        default=5,
+        choices=range(1, cpu_count() + 1),
+        help="Number of workers for multiprocessing",
+    )
+    optional.add_argument(
+        "--token_filename",
+        type=str,
+        help="Filename in home directory that stores token",
+    )
+    optional.add_argument(
+        "--delete_job_dir",
+        action="store_true",
+        help="Delete job directory",
+    )
+
+    args = parser.parse_args()
+
+    (workspace_id, root_dir, kb_base_url, workers, token_filename, delete_job_dir) = (
+        args.workspace_id,
+        args.root_dir,
+        args.kb_base_url,
+        args.workers,
+        args.token_filename,
+        args.delete_job_dir,
+    )
+
+    uid = os.getuid()
+    username = os.getlogin()
+
+    job_dir = _make_job_dir(root_dir, loader_common_names.JOB_DIR, username)
+    output_dir = _make_output_dir(
+        root_dir, loader_common_names.SOURCE_DATA_DIR, SOURCE, workspace_id
+    )
+    boolean_flag = False
+    try:
+        # start podman service
+        proc = loader_helper.start_podman_service(uid)
+        boolean_flag = True
+    except:
+        raise Exception("Podman service fails to start")
+    else:
+        # set up conf and start callback server
+        conf = Conf(job_dir, output_dir, workers, kb_base_url, token_filename)
+
+        visited = set(
+            [
+                upa
+                for upa in os.listdir(output_dir)
+                if loader_helper.is_upa_info_complete(output_dir, upa)
+            ]
+        )
+        print("Skipping: ", visited)
```

Comment:
As the WS dir gets larger and larger this could get more and more expensive. Instead I'd list the objects first, and then check whether they exist already

Reply:
Hi @MrCreosote, 

Let's say output_dir is `/global/cfs/cdirs/kbase/collections/sourcedata/WS/39795` and my current code is gonna loop into each upa to check if upa_info is completed. I may need more info on what you are suggesting. 

<img width="767" alt="Screen Shot 2023-05-02 at 6 51 37 PM" src="https://user-images.githubusercontent.com/22202096/235819464-022b69a0-5329-4beb-bb87-48a3a377235f.png">

Reply:
Resolved for now. Please feel free to reopen.

Reply:
I'm saying list the objects in the workspace first, and then check whether the source directory exists for each object. That way you don't list every directory in the `sourcedata` directory which could be >> than the workspace object list.

Reply:
So it's still listing everything in the directory and then listing objects. Is there a reason you don't want to list objects first and then check if the dir exists for each object as you iterate through the objects?

Reply:
Updated!

Reply:
:+1: